sub-packages/bionemo-dnadl/src/bionemo/dnadl/io/dataset.py

```diff
@@ -0,0 +1,206 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: LicenseRef-Apache2
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: LicenseRef-NvidiaProprietary
+#
+# NVIDIA CORPORATION, its affiliates and licensors retain all intellectual
+# property and proprietary rights in and to this material, related
+# documentation and any modifications thereto. Any use, reproduction,
+# disclosure or distribution of this material and related documentation
+# without an express license agreement from NVIDIA CORPORATION or
+# its affiliates is strictly prohibited.
+
+import os
+from abc import ABC
+from pathlib import Path
+from typing import Any, Callable, List, Optional
+
+import polars as pl
+import pysam
+import torch
+from pyfaidx import Fasta, Sequence
+from torch.utils.data import Dataset
+
+from bionemo.dnadl.tools.genome_editing import create_personal_sequence
+from bionemo.dnadl.tools.genome_interval import GenomeInterval
+from bionemo.dnadl.tools.vcf import SampleId, read_variants_in_interval
+
+
+DNATokenizer = Callable[[str], torch.Tensor]
+
+
+def _thread_safe_genome_access(
+    fasta_file_path: Path, chromosome: str, start_index: int | None = None, end_index: int | None = None
+) -> Sequence:
+    """Bypass the multiprocessing issues with pyfaidx by destroying the fasta object after use."""
+    fasta = Fasta(fasta_file_path)
+    if start_index is not None and end_index is not None:
+        sequence = fasta[chromosome][start_index:end_index]
+    else:
+        sequence = fasta[chromosome]
+    del fasta
```

Comment:
While this may resolve the majority of issues from simultaneous reads, it also introduces a race condition on `.fai` creation and adds a big overhead for reads. I've left a more detailed comment in the main discussion with suggestions.

Reply:
I've now updated to use NvFaidx